build/scripts/script.js

```diff
@@ -519,84 +520,146 @@ function updateSearchResults(value) {
+ * find a match for a h3 section and not for the h1.
+ * To not display h3 orphan in the tree view, we manually add to
+ * the search result the h1 that is a parent to that h3.
+ * @param {array} searchResults
+ * @returns
+ */
+function addMissingTopSections(searchResults) {
+  const h1Sections = {};
+  const missingH1 = [];
+  for (result of searchResults) {
+    if (!h1Sections[result.item.h1]) {
+      h1Sections[result.item.h1] = [];
+    }
+    h1Sections[result.item.h1].push(result);
+  }
+
+  for (const value of Object.values(h1Sections)) {
+    const topSection = value.find((r) => {
+      return r.item.h2 === undefined;
+    });
+    if (topSection === undefined) {
+      const fakeH1 = structuredClone(value[0]);
+      fakeH1.item.h2 = undefined;
+      fakeH1.item.h3 = undefined;
+      fakeH1.item.body = "";
+      missingH1.push(fakeH1);
+    }
+  }
+  return searchResults.concat(missingH1);
+}
+
+/**
+ * Re-orders search results by grouping them according
+ * to their shared section headings (h1, h2, h3).
+ * @param {array} searchResults The array of search results to be re-ordered.
+ * @returns An array re-ordered based on section headings.
+ *
+ * @example
+ */
+function reorderSearchResultByGroup(searchResults) {
+  const results = addMissingTopSections(searchResults);
+
+  // group by h1
+  const groupedSearchResult = groupItems(results);
+
+  return groupedSearchResult;
+}
+
+/**
+ * Groups items in an array based on their h1, h2 and h3 titles.
+ * @param {array} array The array to group.
+ * @returns An object containing items grouped.
+ */
+function groupItems(results) {
+  const grouping = {};
+
+  results.forEach((res) => {
+    let item = res.item;
+
+    if (!grouping[item.file]) {
+      grouping[item.file] = {};
+    }
+    // Create the h1 key
+    if (!grouping[item.file][item.h1]) {
+      grouping[item.file][item.h1] = {};
+    }
+
+    // Create the h2 key, if exists
+    if (item.h2) {
+      if (!grouping[item.file][item.h1][item.h2]) {
+        grouping[item.file][item.h1][item.h2] = {};
       }
-      let linkH3;
-      if (links.anchorH3 !== undefined) {
-        const href = rootUrl + "/" + links.file + "#" + links.anchorH3;
-        linkH3 = document.createElement("a");
-        linkH3.href = href;
+
+      // Create the h3 key, if exists
+      if (item.h3) {
+        if (!grouping[item.file][item.h1][item.h2][item.h3]) {
+          grouping[item.file][item.h1][item.h2][item.h3] = [];
+        }
+        grouping[item.file][item.h1][item.h2][item.h3].push(item);
       } else {
-        linkH3 = document.createElement("span");
+        // If no h3, use '__DEFAULT__'
+        if (!grouping[item.file][item.h1][item.h2]["__DEFAULT__"]) {
+          grouping[item.file][item.h1][item.h2]["__DEFAULT__"] = item;
+        }
+      }
+    } else {
+      // If no h2, use '__DEFAULT__' under h1
+      if (!grouping[item.file][item.h1]["__DEFAULT__"]) {
```

Comment:
As written in another place, I think we shouldn't rely on a string here, as it could break if someone documents something called `__DEFAULT__`

Reply:
I think I fixed it.